Add a new post - Understanding Index Request Cache

kiranprakash154 opened the pull request:
### Description
Add a new post - Understanding Index Request Cache
 
### Issues Resolved
[List any issues this PR will resolve]

### Check List
- [ ] Commits are signed per the DCO using --signoff


By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

kiranprakash154 commented:
(cc: @kkhatua )

kiranprakash154 commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -47,22 +49,26 @@ GET /students/_search?request_cache=true
-1. **CacheEntity** - [IndexShardCacheEntity](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesService.java#L1866C24-L1866C45) that comprises [IndexShard](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java) (think of this as a main ...
-2. **ReaderCacheKeyId** - This is a unique reference to the current state of the shard. On a change of state (i.e., document addition or deletion or updates and upon a refresh) this reference changes.
-3. **BytesReference** - The actual search query in bytes format.
+1. **CacheEntity:** An [IndexShardCacheEntity](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesService.java#L1866C24-L1866C45) that contains the [IndexShard](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java). This reference lin ...
```
An IndexShardCacheEntity that contains the IndexShard seems incorrect, should it be - "An IndexShardCacheEntity contains Index Shard" ?

kolchfa-aws commented:
@pajuric Could you provide meta/keywords for this blog please?

natebower commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+## Cache cleanup
+
+OpenSearch runs a background job called [CacheCleaner](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesService.java#L1678) every minute on a separate thread. This job calls the [`cleanCache`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L698) method, whi ...
```
```suggestion
Every minute, OpenSearch runs a background job called [CacheCleaner](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesService.java#L1678) on a separate thread. This job calls the [`cleanCache`](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L698) method, which iterates through all cache entries, maps each `Key` to a `CleanupKey` in `KeysToClean`, and removes the corresponding entries, as shown in the following diagram.
```

natebower commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+## Wrapping up
+
+Index request cache plays a crucial role in OpenSearch's efficiency. By understanding how it works, you can optimize performance and fine-tune your configurations with greater confidence.
```
```suggestion
The index request cache plays a crucial role in ensuring the efficiency of OpenSearch. By understanding how it works, you can optimize performance and fine-tune your configurations with greater confidence.
```

natebower commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+Index request cache plays a crucial role in OpenSearch's efficiency. By understanding how it works, you can optimize performance and fine-tune your configurations with greater confidence.
+
+OpenSearch thrives on community contributions. If you have ideas or see opportunities for improvement, consider contributing. Your input can help shape the future of this search technology.
```
```suggestion
The OpenSearch Project thrives on community contributions. If you have suggestions for improvement, consider contributing to the project. Your input can help shape the future of this search technology.
```

natebower commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+Index request cache is designed to accelerate search queries in OpenSearch by storing query results at the shard level, as shown in the following image. 
+
+![Index-Request-Cache](/assets/media/blog-images/2024-10-01-understanding-index-request-cache/cache_location.png){:class="img-centered" style="width:750px;"}
```
In this image, please change "Opensearch" to "OpenSearch".

kiranprakash154 replied:
Done, good call !

natebower reviewed (commented):
@kiranprakash154 @pajuric Editorial review complete. Please see my comments and changes and let me know if you have any questions. Thanks!

kiranprakash154 commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+Each cache entry is a key-value pair of **Key → BytesReference**.
+
+A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:
```
I think this should be a `Key` instead of `key` because i want the reader to think of this as a java class which follows the standard of camelcase and its easier to relate with the diagram that also is in the same format.

natebower replied:
@kiranprakash154 Ok, that works for me.

kolchfa-aws replied:
Yes, should be capital. Classes in Java are capitalized.

kiranprakash154 commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+### CleanupKey
+
+When an `IndexReader` is closed, the corresponding `cleanupKey` is added to a set called `keysToClean`, as shown in the following diagram.
+
```
ya good call, its actually referenced in the second image from here. I'll remove `as shown in the following diagram.`

natebower replied:
Ideally, we should leave "as shown in the following diagram" but accurately describe what the diagram shows.

kolchfa-aws replied:
Can we define CleanupKey here as the first sentence? Something like:
```
A CleanupKey corresponds to a Key to be deleted. The relationship between Key and CleanupKey is shown on the following diagram.
<diagram>
When an `IndexReader` is closed, the corresponding `CleanupKey` is added to a set called `KeysToClean`.
```

natebower replied:
"in" the following diagram. Should the instances of "Key" and "CleanupKey" in the first sentence be in code font?

kiranprakash154 replied:
I have changed it to 
"A `CleanupKey` corresponds to a `Key` to be deleted. The relationship between them is shown in the following diagram."

kiranprakash154 commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+A [Key](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesRequestCache.java#L346) consists of three entities:
+
+1. **CacheEntity:** An [IndexShardCacheEntity](https://github.com/opensearch-project/OpenSearch/blob/4199bc2726235456e5b5422eaf4e836f25c2c5ed/server/src/main/java/org/opensearch/indices/IndicesService.java#L1866C24-L1866C45) contains the [IndexShard](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/index/shard/IndexShard.java). This reference links th ...
```
yep, i love that !

kiranprakash154 commented:
@natebower addressed all your suggestions, there is just 1 unresolved comment. [here](https://github.com/opensearch-project/project-website/pull/3361#discussion_r1787447163)

natebower commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,158 @@
+
+### CleanupKey
+
+When an `IndexReader` is closed, the corresponding `CleanupKey` is added to a set called `KeysToClean`
```
```suggestion
When an `IndexReader` is closed, the corresponding `CleanupKey` is added to a set called `KeysToClean`.
```

natebower commented:
> @natebower addressed all your suggestions, there is just 1 unresolved comment. [here](https://github.com/opensearch-project/project-website/pull/3361#discussion_r1787447163)

@kiranprakash154 Thanks! Just one remaining comment for you [here](https://github.com/opensearch-project/project-website/pull/3361/files#r1787465091).

kiranprakash154 commented:
@natebower I have updated the PR with the suggestion and the comment to add intro to the cleanupKey and followed by the diagram.
let me know what u think

natebower reviewed (commented):
Thanks @kiranprakash154 and @kolchfa-aws! LGTM.

Cc: @pajuric

pajuric commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,161 @@
+   - kkhatua
+   - upasagar
+   - kolchfa
+date: 2024-10-01 
```
Update the publish date to today.

pajuric commented on _posts/2024-10-01-understanding-index-request-cache.markdown:
```diff
@@ -0,0 +1,161 @@
+  - technical-posts
+  - search
+meta_keywords: OpenSearch cluster, Opensearch caching, index request cache, search performance optimization, search latency
+meta_description: 
```
Update meta keywords and description with the following: 

_meta_keywords:_   index request cache, search query performance, cache implementation, cache invalidation, cache cleanup, OpenSearch indices request cache

_meta_description:_ Explore the implementation of the OpenSearch index request cache and how this crucial mechanism enhances search query performance and efficiency.

kiranprakash154 commented:
@pajuric pushed the changes u asked.

nateynateynate reviewed (approved):
Looks good!

kiranprakash154 commented:
Thanks @nateynateynate, When will the blog go live ?
